### ISSUE TITLE
Prototype 3-4: PCluster Registration/Admin Demo

### ISSUE THREAD

**Issue by pweil-** [kcp-dev/kcp#418]
### Demo Objective
- [x] Document a repeatable process for registering a physical cluster with kcp
- [x] (stretch goal) Add support in the `kubectl kcp` plugin to register a physical cluster with kcp

### Demo Steps
1. User (or kcp admin, or org admin) needs to add new cluster (capacity, location, whatever)
    1. Choose/Create workspace (this will be the one the syncer will sync).
    2. Create SA in that workspace. The controller automatically generates a token in a `"token"` secret next to the SA.
    3. pcluster admin create kubeconfig with the upper token, and creates that as a secret in the pcluster syncer namespace
1. Reviews the yaml that creates the agent on the cluster
    1.  Sees reasonable roles+rolebindings, a deployment, and has to adapt the kubeconfig secret mount to match secret created above.
    2. YAML contains address of KCP to connect to
    3. YAML contains the unique name of this cluster (according to admin user)
    4. YAML references SA+token to authenticate to KCP, copied from KCP previously
1. Deploys the yaml 
    1. `kubectl apply -f syncer.yaml` 
    2. **Out-of-scope for p3** Possibly more automation around this
    3. Syncer starts by creating a `ClusterWorkspace` CR in KCP
1. Sees the WorkloadCluster show up in the workspace
1. **Out-of-scope for p3**: Does administrative action that makes this available to a workspace
1. Options:
    1. Decommission old cluster
        1. Triggers moves and drains
    1. Add new capacity
        1. The moment the WorkloadCluster is available, the new workload which wasn’t scheduled suddenly gets scheduled and placed (might be a stretch)
    1. Failure
        1. Sees an alert because the cluster is not responding? Don’t have any of our monitoring or conditions done (might be a stretch)

### Action Items

- [x] Scope the current demo as necessary to fit in prototype boundaries
- Technical tasks
   - [x] https://github.com/kcp-dev/kcp/pull/681 wire TokensController into pkg/server/server.go  @sttts  @marun 
   - ~~[ ] create WorkloadCluster object from syncer~~ find and document a procedure to make the syncer work through the service-account @aojea / @marun  
   - [x] build syncer controller that heartbeats @marun
   - [x] #738 build cluster-health controller in the server @imjasonh
   - [x] #723 @aojea 
   - [x] minimal, experimental/temporary RBAC policy on the pcluster
   - ~~[ ] **stretch-goal:** privileged syncer-namespace-controller and non-admin pcluster permissions for syncer~~https://github.com/kcp-dev/kcp/issues/387
   - [x] Out-of-scope (prototype 4 or 5): kcp plugin
   - Out-of-scope (prototype 4): any kind of in-cluster object with the resource list to sync
   - Out-of-scope (prototype 4): multi-workspace support (no "*" informers of any kind)
   - Out-of-scope: node-authorizer-like authorizer in kcp (probably replaced by virtual workspace)
   - Out-of-scope: lease management to avoid two syncer with the same cluster name
   - Out-of-scope: anything about bootstrap tokens, CSR, ...
- [x] ~~Contribute to final demo script and recording~~ part of @ncdc's youtube video https://www.youtube.com/watch?v=0trUxoQqdNk&t=1s
- ~~[ ] Remove KubeConfig from Cluster.Spec (TODO file issue for this)~~ https://github.com/kcp-dev/kcp/issues/934

**Comment by stevekuznetsov** [comment]
Do we want to commit to 5 here being the addition of ThingClusterCapacityChunk to a ThingLocation?

**Comment by stevekuznetsov** [comment]
This is transparent multi-cluster  ... admin was supposed to be about movement of data _between kcp instances in a sharded fleet_ but we are no longer targeting that deployment modality for the forseeable future. What value does this demo have on top of TMC?

**Comment by pweil-** [comment]
> admin was supposed to be about ...

After we scoped out sharding this is where the conversation went on what can be shown from an admin perspective.  In terms of actual work, if we agree that it is simply to create the demo once the TMC work is done then I think we're good here.

**Comment by imjasonh** [comment]
+1 to Steve's point, this script sounds more like a demo of pcluster registration and administration, not about "control plane resiliency, security, defense in-depth"

The only problem there is, I don't think anybody's current working on designing/prototyping pcluster registration, so if this is part of the demo it's unassigned.

**Comment by pweil-** [comment]
> The only problem there is, I don't think anybody's current working on designing/prototyping pcluster registration, so if this is part of the demo it's unassigned.

That puts us in the same boat.  We can either start scoping those tasks out or choose to call this duplicate (or just a different recording where the driver is "admin" and not "user").  Since pcluster registration doesn't push us any further on the demo objectives and scope is already pretty large for the remaining time I'd probably vote for the latter (and leave whether or not we want a different recording up to further discussion).

**Comment by jmprusi** [comment]
> +1 to Steve's point, this script sounds more like a demo of pcluster registration and administration, not about "control plane resiliency, security, defense in-depth"
> 
> The only problem there is, I don't think anybody's current working on designing/prototyping pcluster registration, so if this is part of the demo it's unassigned.

I'm working on that currently and tried to doc some of those topics here: https://docs.google.com/document/d/1pDv2OVXpT5TUgO0rKNFwsht1EFasUE_15mvvBY_R-LA/edit#

Let's have a chat and try to create some issues?

**Comment by imjasonh** [comment]
For cluster registration we need to produce:
- YAML describing a syncer deployment that can be applied to a pcluster
- a working demo, even better a script, ideally a `kubectl` plugin surface that:
  - creates a SA in kcp
  - sees a SA token secret created for that SA, associated with that SA
  - copies that SA token to the pcluster, associated with the SA running the syncer in the pcluster
  - ...so that syncer has all the auth it needs to talk to kcp
- RBAC to configure in kcp to correctly(/minimally) scope permissions for each pcluster's syncer SA
  - e.g., create Cluster objects corresponding to that pcluster syncer installation

...further processes may automate this very manual process even more, possibly outside the scope of kcp itself (TBD)

For cluster health-checking / heartbeats, syncer should use configuration (from above ☝️) to periodically update its Cluster resource in kcp to say it's healthy
- kcp needs to take that healthcheck into account when determining cluster `Ready` condition, which triggers scheduling decisions, etc.

@jmprusi @ncdc anything I missed?

**Comment by ncdc** [comment]
👍

Also talked to @sttts @s-urbaniak and 👍 for enabling the kube `TokensController` and a corresponding authorizer.

Also need to remove the kubeconfig field from the `Cluster` CRD

**Comment by sttts** [comment]
Marked anything about Location and "making cluster available in a workspace" as out-of-scope for p3.

**Comment by marun** [comment]
> create WorkloadCluster object from syncer @aojea

I just noticed this work item.  What is the rational for the syncer being responsible for creating a WorkloadCluster resource vs requiring that it exist already?

**Comment by sttts** [comment]
> I just noticed this work item. What is the rational for the syncer being responsible for creating a WorkloadCluster resource vs requiring that it exist already?

I don't think this is set in stone. The syncer will write the status. So it actually would be odd if it also wrote the spec (which it would on create).

In the future we will probably have some tool or even a controller preparing a kubeconfig, service account, roles, role binding. That tool or controller then also create the WorkloadCluster object. Sounds fine.

For the issue here: we need a kcp RBAC policy for the syncer service account.

**Comment by ncdc** [comment]
Moving to 0.4. When we next review issues in the milestone, we should either close this (if it's done) or update w/what's left to do.
